### ISSUE TITLE
Inconsistent/incorrect handling of MKL and `MKL_INT`

### ISSUE THREAD

**Issue by cwpearson** [kokkos/kokkos-kernels#2060]
The following is my current understanding...

[Intel ships an "ILP64" and an "LP64" flavor of libraries](https://www.smcm.iqfr.csic.es/docs/intel/mkl/mkl_userguide/MKL_UG_linking_your_application/Support_for_ILP64_Programming.htm), e.g. `libmkl_intel_lp64.so` vs `libmkl_intel_ilp64.so`.

* ILP64: 64-bit integer index types
* LP64: 32-bit integer index types

At config time, Kokkos Kernels will print something like this:
```bash
...
-- Found MKL: /projects/x86-64-icelake-rocky8/tpls/intel-oneapi-mkl/2023.1.0/oneapi/2023.1.0/base/s4p2s4d/lib/intel64/libmkl_intel_ilp64.so  
...
```
in this example, linking against ilp64.

Since both types are supported, MKL uses an `MKL_INT` (and friends) type, which should be either a 32 or 64-bit integer type, accordingly. The user must match up the `MKL_INT` they want with the libraries they're linking against.

Sadly, our HostBlas interface expects things to be `int*`, for example, `m`, `n`, `k`, `lda`, etc in gemm. This means when using MKL in ILP64 mode, we are passing an `int*` to a place where MKL wants an `int64_t *`, causing many and sundry (and not always loud) errors.

I think the overall solution to this is in HostBlas, we need to use a new `KK_INT` type wherever we current use `int`, and for the case of Intel, or other folks with 64-bit BLAS extensions, we'll define `KK_INT` appropriately.

We may also need some process for choosing ilp64 vs lp64 when linking, or just link both?

Previous work along these lines:
* https://github.com/kokkos/kokkos-kernels/pull/1868 to at least partly handle this for spmv and spgemm
* https://github.com/kokkos/kokkos-kernels/issues/216 is an issue just in a performance test

**Comment by lucbv** [comment]
In the case of PR #1868 only sparse functions were modified to fix the MKL_INT issue, this was to an extent simpler since in that case there is not support for a standard host TPL. In the case of BLAS functions however we would like to keep the generic approach of `KokkosBlas_Host_tpl.{hpp,cpp}` if possible.
I agree with the KK_INT approach, that is probably the simplest way to handle the issue, annoyingly we probably want to carefully read the whole MKL documentation to make sure they consistently use MKL_INT and do not have any hardcoded `int` anywhere : (

**Comment by lucbv** [comment]
This was fixed with PR #2110 so closing this now.
